### PR TITLE
improve error messages when running plan without parameters

### LINKED ISSUE #206: Unclear error message when running plan without parameters
When the controller run subcommand is executed without providing any parameters (`-p`), the following is returned:
```
$ blueapi controller run scan
Traceback (most recent call last):
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/bin/blueapi", line 8, in <module>
    sys.exit(main())
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/src/blueapi/cli/cli.py", line 67, in wrapper
    func(*args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/venv/lib/python3.10/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/scratch/30day_tmp/delete15may/bs/blueapi/src/blueapi/cli/cli.py", line 106, in run_plan
    json=json.loads(parameters),
  File "/dls_sw/apps/python/miniforge/4.10.0-0/envs/python3.10/lib/python3.10/json/__init__.py", line 339, in loads
    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
TypeError: the JSON object must be str, bytes or bytearray, not NoneType
```

This is not particularly user friendly, if parameters are always required, can this be made into a mandatory argument?

### PR DESCRIPTION
**Pull request by stan-dot**
closes #206

### CODE REVIEW

**Comment by stan-dot** [comment]
one error in the `test_cli.py`. the test must be updated to first expect the helper call to the API


https://github.com/DiamondLightSource/blueapi/actions/runs/8192137690/job/22402991512?pr=385

File: .vscode/settings.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "python.languageServer": "Pylance",
     "editor.formatOnSave": true,
     "editor.codeActionsOnSave": {
-        "source.organizeImports": true
+        "source.organizeImports": "explicit"
```
Comment (joeshannon) [inline .vscode/settings.json:13]: Is this change intended?

File: src/blueapi/cli/cli.py [hunk 2]
```diff
@@ -182,6 +183,22 @@ def store_finished_event(event: WorkerEvent) -> None:
             finished_event.append(event)
 
     parameters = parameters or "{}"
+    schema: PlanModel = client.get_plan(name)
+    progress_tracking = f"Trying to run plan: {name} with the expected schema: {schema}"
+    print(progress_tracking)
```
Comment (joeshannon) [inline src/blueapi/cli/cli.py:188]: I'm not very familiar with using the CLI and how noisy/verbose it is but this function is now printing quite a lot of output which is arguably debug/verbose info.

**Review by joeshannon — COMMENTED** [review]
I haven't tested actually using it yet but will aim to tomorrow.

Two of the tests are currently failing though, is this expected?

File: src/blueapi/cli/cli.py (hunk 2 above)
Comment (stan-dot) [reply src/blueapi/cli/cli.py:188]: got a variation of the latter

File: .vscode/settings.json (hunk 1 above)
Comment (stan-dot) [reply .vscode/settings.json]: fixed this now

**Review by callumforrester — APPROVED** [review]

File: src/blueapi/cli/cli.py (hunk 2 above)
Comment (stan-dot) [reply src/blueapi/cli/cli.py:188]: I already got this highlight of what does not match

Comment (stan-dot) [reply src/blueapi/cli/cli.py:188]: image
the latest branch looks diff

Comment (stan-dot) [reply src/blueapi/cli/cli.py:188]: I don't think a new ticket is needed at this point

**Comment by stan-dot** [comment]
fwiw the REST part of CLI has `create_task` while the server has `submit_task`. should this naming be consistent? @callumforrester

**Review by joeshannon — CHANGES_REQUESTED** [review]
It's been quite a while since the issue was opened and things have changed quite a bit now.


The approach used looks reasonable, and does prevent the very nasty server/pool errors, however running `blueapi controller run scan` with this PR gives:

```
Traceback (most recent call last):
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/bin/blueapi", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/src/blueapi/cli/cli.py", line 108, in wrapper
    func(*args, **kwargs)
  File "/scratch/30day_tmp/delete24apr/blueapi/venv/lib/python3.11/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/30day_tmp/delete24apr/blueapi/src/blueapi/cli/cli.py", line 188, in run_plan
    task = Task(name=name, params=parsed_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Task
params
  value is not a valid dict (type=type_error.dict)
```
which doesn't seem to be an improvement over the original error reported in the issue.

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+    ],
     "python.testing.unittestEnabled": false,
     "python.testing.pytestEnabled": true,
     "editor.formatOnSave": true,
     "editor.codeActionsOnSave": {
         "source.organizeImports": "explicit"
     },
     "[python]": {
-        "editor.defaultFormatter": "charliermarsh.ruff",
+        "editor.defaultFormatter": "ms-python.black-formatter",
```
[the remark this answers is not in the document]
Comment (stan-dot) [reply .vscode/settings.json:16]: thanks for feedback @joeshannon , will push a commit responding today

File: src/blueapi/cli/cli.py [hunk 4]
```diff
@@ -180,8 +182,10 @@ def store_finished_event(event: WorkerEvent) -> None:
         if event.is_complete():
             finished_event.append(event)
 
-    parameters = parameters or "{}"
-    task = Task(name=name, params=json.loads(parameters))
+    parsed_params: Mapping[str, Any] = (
```
[the remark this answers is not in the document]
Comment (stan-dot) [reply src/blueapi/cli/cli.py:185]: I added the check because the linter complained. I don't see `json.loads` inside Task constructor, as a function it expects Mapping[str, Any] while the original parameters is a str.

File: src/blueapi/cli/rest.py [hunk 5]
```diff
@@ -105,9 +115,17 @@ def _request_and_deserialize(
...
     ) -> T:
         url = self._url(suffix)
+        # todo consider changing this to a switch statement for easier mocking
+        # alternatively do single responsiblity principle
+        # and decoule request from deserializing.
         response = requests.request(method, url, json=data)
         if raise_if(response):
-            raise BlueskyRemoteError(str(response))
+            message = get_status_message(response.status_code)
+            t = response.text
```
[the remark this answers is not in the document]
Comment (stan-dot) [reply src/blueapi/cli/rest.py:124]: right

**Comment by stan-dot** [comment]
in the CLI file we don't have access to the raw http response as that's in the `request_and_deserialize` file.

**Comment by stan-dot** [comment]
> `request_and_deserialize`

that is a too optmistic a function, does not provide natural thin error handling from the http to cli layer

**Comment by stan-dot** [comment]
suggested plan of action - hard code the request into this specific cli command - without using the `request_and_deserialize`

**Comment by callumforrester** [comment]
@stan-dot You can do what you like within the `RestClient` class, but I don't think the HTTP response should be exposed outside

**Comment by stan-dot** [comment]
the thing is that right now `_request_and_deserialize` is defined to necessarily return a TargetType. but this function by name handles 2 things that do not necessarily follow one another. so this function if wants to join those 2 things will necessarily have handling of a rejected request. But we don't want to terminate the operation there. we want to provide output in the cli. 

we can throw the error, but then the function does not return normally. 

one alternative is to change the return tpe from 'T' to `tuple[T |None, E| None]`.
the second alternative is to just throw an error in the RestClient and hope that the cli will catch it. 

NOTE: if auto-client gen succeeds then this is a moot question

**Comment by stan-dot** [comment]
that's a better serverside message but we don't want a 5** message but a 4** one. on it

>  blueapi controller run scan '{"tim":5}'
('server error with this message: Response failed with text: Internal Server '
 'Error,\n'
 '            with error code: 500\n'
 '            which corresponds to Internal Server Error ')
 
 is the behavior

**Comment by stan-dot** [comment]
image
how is this supposed to handle errors ???? there is a fundamental architectural void in the design of how it is now.


suggestion: do client autogen issue first #364

**Comment by stan-dot** [comment]
doing this fresh
